src/fundus/scraping/common_crawl/pipeline.py

```diff
@@ -0,0 +1,284 @@
+                with None values for respective attributes in the extraction or by skipping entire articles.
+                If set to "catch", errors will be caught as attribute values or, if an entire article fails,
+                through Article.exception. If set to "raise", all errors encountered during extraction will
+                be raised. Defaults to "suppress".
+            only_complete (Union[bool, ExtractionFilter]): Set a callable satisfying the ExtractionFilter
+                protocol as an extraction filter or use a boolean. If False, all articles will be yielded,
+                if True, only those with all attributes extracted. Defaults to ExtractionFilter letting
+                through all articles with at least title, body, and publishing_date set.
+            url_filter (Optional[URLFilter]): A callable object satisfying the URLFilter protocol to skip
+                URLs before download. This filter applies on both requested and responded URL. Defaults to None.
+            only_unique (bool): If set to True, articles yielded will be unique on the responded URL.
+                Always returns the first encountered article. Defaults to True.
+
+        Returns:
+            Iterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        if max_articles == 0:
+            return
+
+        if max_articles is None:
+            max_articles = -1
+
+        def build_extraction_filter() -> Optional[ExtractionFilter]:
+            if isinstance(only_complete, bool):
+                return (
+                    None
+                    if only_complete is False
+                    else lambda extracted: not all(
+                        bool(v) if not isinstance(v, Exception) else False for _, v in extracted.items()
+                    )
+                )
+            else:
+                return only_complete
+
+        warc_paths = self._get_warc_paths(start, end)
+        response_cache: Set[str] = set()
+
+        article_task: Callable[[str], Iterator[Article]] = partial(
+            self._fetch_articles,
+            publishers=self.publishers,
+            error_handling=error_handling,
+            extraction_filter=build_extraction_filter(),
+            url_filter=url_filter,
+        )
+
+        if self.processes == 0:
+            article_iter = self._single_crawl(warc_paths, article_task)
+        else:
+            article_iter = self._parallel_crawl(warc_paths, article_task)
+
+        for article_idx, article in enumerate(article_iter, start=1):
+            if not only_unique or article.html.responded_url not in response_cache:
+                response_cache.add(article.html.responded_url)
+                yield article
+            if article_idx == max_articles:
+                break
+
+
+def pool_queue_iter(handle: MapResult[Any], queue: Queue[_T]) -> Iterator[_T]:
+    while True:
```

Comment:
Adds the docstring from the previous class to the refactored function.
Also, I think it would make sense to move this function to the top of the file with all the other utility functions.

```suggestion
def pool_queue_iter(handle: MapResult[Any], queue: Queue[_T]) -> Iterator[_T]:
    """Utility function to iterate exhaustively over a pool queue.

    The underlying iterator of this function repeatedly exhausts the given queue.
    Then, if the queue is empty only if all the pool's jobs have finished, the iterator reruns.
    Otherwise, it waits for the queue to be populated with the next result from the pool.

    Args:
        handle (MapResult[Any]):  A handle o the MappedResult of the underling multiprocessing pool.
        queue (Queue[_T]): The pool queue.

    Returns:
        Iterator[_T]: The iterator over the queue as it is populated.
    """
    while True:
```